lib/HeroBonus.cpp

```diff
@@ -1840,7 +1842,49 @@ int LimiterList::limit( const BonusLimitationContext &context ) const
 	return wasntSure ? ILimiter::NOT_SURE : ILimiter::ACCEPT;
 }
 
-void LimiterList::add( TLimiterPtr limiter )
+void AllOfLimiter::add(TLimiterPtr limiter)
+{
+	limiters.push_back(limiter);
+}
+
+int AnyOfLimiter::limit(const BonusLimitationContext & context) const

```

Comment:
All that `limit` functions duplicate each other at 99% and (important) differences are not that easy to see. You should merge "logical" limiters into one class